### ISSUE TITLE
thread 'main' panicked at /home/appuser/.cargo/git/checkouts/sp1-20c98843a1ffc860/8ee9a84/recursion/gnark-ffi/src/groth16.rs:382:10

### ISSUE THREAD

**Issue by gusgkr0117** [succinctlabs/sp1-project-template#2]
I run the source code under a docker image using Dockerfile below.
```
FROM ubuntu:22.04@sha256:bcc511d82482900604524a8e8d64bf4c53b2461868dac55f4d04d660e61983cb

RUN apt-get update
RUN apt-get install -y --no-install-recommends ca-certificates clang curl libssl-dev pkg-config git dialog
RUN groupadd --gid 1000 appuser
RUN useradd -s /bin/bash -u 1000 -g appuser -m appuser
RUN apt-get install -y sudo
RUN echo appuser ALL=\(ALL\) NOPASSWD: ALL >> /etc/sudoers.d/appuser
RUN chmod 0440 /etc/sudoers.d/appuser
USER appuser

RUN curl --proto '=https' --tlsv1.2 --retry 10 --retry-connrefused -fsSL 'https://sh.rustup.rs' | sh -s -- -y
ENV PATH="/home/appuser/.cargo/bin:${PATH}"
RUN curl -L https://sp1.succinct.xyz | bash && ~/.sp1/bin/sp1up

WORKDIR /home/appuser/program

ENV CARGO_TERM_COLOR=always
ENTRYPOINT ["/bin/bash"]
# ENTRYPOINT [ "/root/.sp1/bin/cargo-prove" ]
# CMD [ "prove" ]
```

After creating a container using above docker image, I cloned this `sp1-project-template` repository and run the command below.
```
RUST_LOG=info cargo run --package fibonacci-script --bin prove --release
```

Then I got following error messages.
```
warning: skipping duplicate package `fibonacci-program` found at `/home/appuser/.cargo/git/checkouts/sp1-20c98843a1ffc860/8ee9a84/examples/fibonacci/program`
warning: fibonacci-script@0.1.0: fibonacci-program built at 2024-05-10 05:04:59
    Finished `release` profile [optimized] target(s) in 0.29s
     Running `target/release/prove`
[sp1]: groth16 artifacts already seem to exist at /home/appuser/.sp1/circuits/49fedbe4. if you want to re-download them, delete the directory
2024-05-10T06:50:38.760285Z  INFO initialize prover: recursion program size: 2367
2024-05-10T06:50:40.677075Z  INFO initialize prover: recursion program size: 988878
2024-05-10T06:50:42.075088Z  INFO initialize prover:setup machine: close time.busy=1.35s time.idle=2.82µs
2024-05-10T06:50:46.031347Z  INFO initialize prover:setup machine: close time.busy=3.96s time.idle=2.41µs
2024-05-10T06:50:56.387934Z  INFO initialize prover:setup machine: close time.busy=10.4s time.idle=2.27µs
2024-05-10T06:50:56.493463Z  INFO initialize prover: close time.busy=17.8s time.idle=3.09µs
2024-05-10T06:50:56.688459Z  INFO setup:setup machine: close time.busy=194ms time.idle=1.98µs
2024-05-10T06:50:56.688714Z  INFO setup: close time.busy=195ms time.idle=1.23µs
2024-05-10T06:50:57.014704Z  INFO prove_core:setup machine: close time.busy=223ms time.idle=3.54µs
2024-05-10T06:50:57.014754Z  INFO prove_core: loading memory image
2024-05-10T06:50:57.015433Z  INFO prove_core: starting execution
2024-05-10T06:50:57.015445Z  INFO prove_core: clk = 0 pc = 0x20204c    
stdout: n: 500
stdout: a: 1268
stdout: b: 1926
2024-05-10T06:50:57.019053Z  INFO prove_core: finished execution clk = 18653 pc = 0x0
2024-05-10T06:50:57.312829Z  INFO prove_core:setup machine: close time.busy=191ms time.idle=3.26µs
2024-05-10T06:50:57.340103Z  INFO prove_core:shard_record: bn254_add_events = 0    
2024-05-10T06:50:57.340133Z  INFO prove_core:shard_record: shift_right_events = 2395    
2024-05-10T06:50:57.340136Z  INFO prove_core:shard_record: ed_add_events = 0    
2024-05-10T06:50:57.340139Z  INFO prove_core:shard_record: bitwise_events = 3342    
2024-05-10T06:50:57.340144Z  INFO prove_core:shard_record: blake3_compress_inner_events = 0    
2024-05-10T06:50:57.340147Z  INFO prove_core:shard_record: bls12381_double_events = 0    
2024-05-10T06:50:57.340149Z  INFO prove_core:shard_record: sha_extend_events = 0    
2024-05-10T06:50:57.340153Z  INFO prove_core:shard_record: secp256k1_add_events = 0    
2024-05-10T06:50:57.340155Z  INFO prove_core:shard_record: sub_events = 1096    
2024-05-10T06:50:57.340159Z  INFO prove_core:shard_record: add_events = 9087    
2024-05-10T06:50:57.340161Z  INFO prove_core:shard_record: uint256_mul_events = 0    
2024-05-10T06:50:57.340163Z  INFO prove_core:shard_record: keccak_permute_events = 0    
2024-05-10T06:50:57.340167Z  INFO prove_core:shard_record: secp256k1_double_events = 0    
2024-05-10T06:50:57.340169Z  INFO prove_core:shard_record: bn254_double_events = 0    
2024-05-10T06:50:57.340173Z  INFO prove_core:shard_record: k256_decompress_events = 0    
2024-05-10T06:50:57.340177Z  INFO prove_core:shard_record: ed_decompress_events = 0    
2024-05-10T06:50:57.340179Z  INFO prove_core:shard_record: bls12381_add_events = 0    
2024-05-10T06:50:57.340181Z  INFO prove_core:shard_record: bls12381_decompress_events = 0    
2024-05-10T06:50:57.340184Z  INFO prove_core:shard_record: sha_compress_events = 0    
2024-05-10T06:50:57.340189Z  INFO prove_core:shard_record: shift_left_events = 1330    
2024-05-10T06:50:57.340191Z  INFO prove_core:shard_record: lt_events = 2022    
2024-05-10T06:50:57.340193Z  INFO prove_core:shard_record: cpu_events = 18653    
2024-05-10T06:50:57.340196Z  INFO prove_core:shard_record: divrem_events = 1    
2024-05-10T06:50:57.340198Z  INFO prove_core:shard_record: mul_events = 1046    
2024-05-10T06:50:57.341139Z  INFO prove_core:shard_record: close time.busy=28.3ms time.idle=1.37µs
2024-05-10T06:50:58.874995Z  INFO prove_core:prove_shards: close time.busy=1.53s time.idle=1.43µs
2024-05-10T06:50:58.876168Z  INFO prove_core: summary: cycles=18653, e2e=1562, khz=11.94, proofSize=1.09 MiB
2024-05-10T06:50:58.879455Z  INFO prove_core: close time.busy=2.19s time.idle=1.34µs
2024-05-10T06:51:00.466350Z  INFO compress:verify_batch: runtime summary: cycles=2604151, nb_poseidons=36235
2024-05-10T06:51:00.842912Z  INFO compress:verify_batch:shard_record: poseidon2_events = 36235    
2024-05-10T06:51:00.842946Z  INFO compress:verify_batch:shard_record: cpu_events = 2604151    
2024-05-10T06:51:00.842950Z  INFO compress:verify_batch:shard_record: range_check_events = 47595    
2024-05-10T06:51:00.842953Z  INFO compress:verify_batch:shard_record: fri_fold_events = 175800    
2024-05-10T06:51:00.842957Z  INFO compress:verify_batch:shard_record: close time.busy=58.0µs time.idle=2.73µs
2024-05-10T06:51:41.324205Z  INFO compress:verify_batch:prove_shards: close time.busy=40.5s time.idle=1.83µs
2024-05-10T06:51:41.326057Z  INFO compress:verify_batch: proving summary: cycles=2604151, e2e=40.857899211, khz=63.74, proofSize=1.59 MiB
2024-05-10T06:51:41.412945Z  INFO compress:verify_batch: close time.busy=42.5s time.idle=1.84µs
2024-05-10T06:51:41.413470Z  INFO compress: close time.busy=42.5s time.idle=1.05µs
2024-05-10T06:51:43.110008Z  INFO shrink:verify_batch: runtime summary: cycles=2542907, nb_poseidons=51653
2024-05-10T06:51:43.499208Z  INFO shrink:verify_batch:shard_record: range_check_events = 41684    
2024-05-10T06:51:43.499242Z  INFO shrink:verify_batch:shard_record: cpu_events = 2542907    
2024-05-10T06:51:43.499245Z  INFO shrink:verify_batch:shard_record: poseidon2_events = 51653    
2024-05-10T06:51:43.499248Z  INFO shrink:verify_batch:shard_record: fri_fold_events = 218000    
2024-05-10T06:51:43.499252Z  INFO shrink:verify_batch:shard_record: close time.busy=57.1µs time.idle=2.97µs
2024-05-10T06:53:33.566534Z  INFO shrink:verify_batch:prove_shards: close time.busy=110s time.idle=1.40µs
2024-05-10T06:53:33.567260Z  INFO shrink:verify_batch: proving summary: cycles=2542907, e2e=110.456560284, khz=23.02, proofSize=566 KiB
2024-05-10T06:53:33.649304Z  INFO shrink:verify_batch: close time.busy=112s time.idle=1.00µs
2024-05-10T06:53:33.649337Z  INFO shrink: close time.busy=112s time.idle=1.30µs
2024-05-10T06:53:34.131973Z  INFO wrap_bn254:verify_batch: runtime summary: cycles=893258, nb_poseidons=17638
2024-05-10T06:53:34.252559Z  INFO wrap_bn254:verify_batch:shard_record: fri_fold_events = 50754    
2024-05-10T06:53:34.252597Z  INFO wrap_bn254:verify_batch:shard_record: range_check_events = 47855    
2024-05-10T06:53:34.252600Z  INFO wrap_bn254:verify_batch:shard_record: poseidon2_events = 17638    
2024-05-10T06:53:34.252603Z  INFO wrap_bn254:verify_batch:shard_record: cpu_events = 893258    
2024-05-10T06:53:34.252608Z  INFO wrap_bn254:verify_batch:shard_record: close time.busy=61.7µs time.idle=2.60µs
2024-05-10T06:55:22.714096Z  INFO wrap_bn254:verify_batch:prove_shards: close time.busy=108s time.idle=1.33µs
2024-05-10T06:55:22.717884Z  INFO wrap_bn254:verify_batch: proving summary: cycles=893258, e2e=108.582143474, khz=8.23, proofSize=749 KiB
2024-05-10T06:55:22.733797Z  INFO wrap_bn254:verify_batch: close time.busy=109s time.idle=1.22µs
2024-05-10T06:55:22.733825Z  INFO wrap_bn254: close time.busy=109s time.idle=1.19µs
[sp1]: groth16 artifacts already seem to exist at /home/appuser/.sp1/circuits/49fedbe4. if you want to re-download them, delete the directory
thread 'main' panicked at /home/appuser/.cargo/git/checkouts/sp1-20c98843a1ffc860/8ee9a84/recursion/gnark-ffi/src/groth16.rs:382:10:
called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
2024-05-10T06:55:22.734831Z  INFO wrap_groth16: close time.busy=773µs time.idle=1.26µs
```

How can I fix this error?

**Comment by pedrohba1** [comment]
I've attempted to use the Dockerfile to run the script and I ended up in the same error as you.

**Comment by gusgkr0117** [comment]
I was able to run the script successfully in my ubuntu vm image. So I guess that groth16 artifacts are not downloaded completely in the docker container. it seems that there is a limit of storage size of the container and it is causing error(i'm not sure). but I haven't searched any solutions yet.

**Comment by ratankaliani** [comment]
Hey @pedrohba1 @gusgkr0117, thanks for raising this issue! 

We resolved this in a recent release of this repo, so if you re-pull this example, it should work with no issues. 

For a bit of context on the error, in versions of SP1 <= `v1.0.7-testnet`, there was a requirement that Go was installed in order to verify & generate the PLONK proofs locally. Now, the local proof generation & verification occurs inside of a Docker container by default so that users don't need to have Go installed in order to compile SP1. 

The issue that you saw in the previous version of this repo was due to the fact your Docker container did not have Go installed. Given that you are running the example in a Docker container, I'd recommend installing Go and activating the `native-gnark` feature on the `sp1-sdk`, which will handle the proof generation & verification using Go bindings, instead of spinning up a Docker container inside of your Docker container.

Keep in mind the warnings in https://github.com/succinctlabs/sp1-project-template?tab=readme-ov-file#standard-proof-generation about the minimum memory requirements for proof generation.
